spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIcebergProcedures.scala

```diff
@@ -0,0 +1,132 @@
+import org.apache.spark.sql.catalyst.expressions.{Expression, Literal}
+import org.apache.spark.sql.catalyst.plans.logical.{Call, CallArgument, LogicalPlan}
+import org.apache.spark.sql.catalyst.procedures.{IcebergProcedureRegistry, OptionalParameter, Parameter, Procedure}
+import org.apache.spark.sql.catalyst.rules.Rule
+import scala.collection.mutable
+
+// TODO: case sensitivity?
+object ResolveIcebergProcedures extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case Call(nameParts, args) =>
+      val procedure = IcebergProcedureRegistry.resolve(nameParts)
+      val values = buildProcedureValues(procedure, args)
+      procedure.createRunnableCommand(values)
+  }
+
+  private def buildProcedureValues(procedure: Procedure, args: Seq[CallArgument]): Array[Any] = {
+    // build a map of declared parameter names to their positions
+    val params = procedure.parameters
+    val nameToPositionMap = params.map(_.name).zipWithIndex.toMap
+
+    // verify named and positional args are not mixed
+    val containsNamedArg = args.exists(arg => arg.name.isDefined)
+    val containsPositionalArg = args.exists(arg => arg.name.isEmpty)
+    if (containsNamedArg && containsPositionalArg) {
+      throw new AnalysisException("Named and positional arguments cannot be mixed")
+    }
+
+    // build a map of parameter names to args
+    val nameToArgMap = if (containsNamedArg) {
+      buildNameToArgMapUsingNames(args, nameToPositionMap)
+    } else {
+      buildNameToArgMapUsingPositions(args, params)
+    }
+
+    // verify all required parameters are provided
+    params.filter(_.required)
+      .find(param => !nameToArgMap.contains(param.name))
+      .foreach { missingArg =>
+        throw new AnalysisException(s"Required procedure argument '${missingArg.name}' is missing")
+      }
+
+    val values = new Array[Any](params.size)
+
+    // convert provided args from internal Spark representation to Scala
+    nameToArgMap.foreach { case (name, arg) =>
+      val position = nameToPositionMap(name)
+      val param = params(position)
+      val paramType = param.dataType
+      val argType = arg.expr.dataType
+      if (paramType != argType) {
+        throw new AnalysisException(s"Wrong arg type for '${param.name}': expected $paramType but got $argType")
+      }
+      values(position) = toScalaValue(arg.expr)
+    }
+
+    // assign default values for optional params
+    params.foreach {
+      case p: OptionalParameter if !nameToArgMap.contains(p.name) =>
+        val position = nameToPositionMap(p.name)
+        values(position) = p.defaultValue
+      case _ =>
+    }
+
+    values
+  }
+
+  private def buildNameToArgMapUsingNames(
+      args: Seq[CallArgument],
+      nameToPositionMap: Map[String, Int]): Map[String, CallArgument] = {
+
+    val nameToArgMap = mutable.LinkedHashMap.empty[String, CallArgument]
+    args.foreach { arg =>
```

Comment:
I tend to like doing
args.map { x => (k,v) }.toMap

Especially if we are going to fail on any bad option, no need to use the mutable Map

Reply:
Ah the duplicate check makes this less possible :/ 
I could say this could be fold then but that's probably not necessary either ...

Reply:
Actually maybe there should be a separate validateArgs function that you can run before building the name map ... just a thought

Reply:
Will reconsider this next.

Reply:
@RussellSpitzer we could have this:

```
    val namedArgs = args.asInstanceOf[Seq[NamedArgument]]

    namedArgs.groupBy(_.name)
      .foreach { case (name, matchingArgs) =>
        if (matchingArgs.size > 1) {
          throw new AnalysisException(s"Duplicate procedure argument: '$name'")
        }
        if (!nameToPositionMap.contains(name)) {
          throw new AnalysisException(s"Unknown argument name: '$name'")
        }
      }
    
    namedArgs.map(arg => arg.name -> arg).toMap
```

But `groupBy(_.name)` will use a mutable map under the hood. Plus, the main motiviation of using `LinkedHashMap` was to preserve the order. Later, we iterate through this map and do some validation and it would probably make sense to validate the args in the order they are defined (probably not a big deal?). Writing a recursive function and building an immutable set would probably make things more complicated that it should.

I am flexible here.

Reply:
Mostly I don't want us to make our own maps explicitly, I think the above is fine, but I think if we want to be really friendly, we gather up all the validation errors at once. 

Perhaps something like
```scala
val validationErrors = NamedArgs.groupBy(_.name).collect {
   case (name, matchingArgs) if matchinArgs.size >1 => s"Duplicate Procedure argument: '$name'") // May want to add values here as well
   case (name, matchingArgs) if !nameToPositionMap.contains(name) => s"Unknown argument name: '$name'")

if (validationErrors.nonEmpty) {
   throw new AnalysisException(s"Found the following parameter violations: validationErrors.join)
}
```

Also I love having my if statements in my case statements :)

Reply:
Agree, I think we better indicate all vioalations. I'll rework this part.